mlir/include/mlir/Rewrite/PatternApplicator.h

```diff
@@ -68,9 +68,9 @@ class PatternApplicator {
   ///            invalidate the match and try another pattern.
   LogicalResult
   matchAndRewrite(Operation *op, PatternRewriter &rewriter,
-                  function_ref<bool(const Pattern &)> canApply = {},
-                  function_ref<void(const Pattern &)> onFailure = {},
-                  function_ref<LogicalResult(const Pattern &)> onSuccess = {});
+                  std::function<bool(const Pattern &)> canApply = {},
+                  std::function<void(const Pattern &)> onFailure = {},
+                  std::function<LogicalResult(const Pattern &)> onSuccess = {});
```

Comment:
Why this change?

Reply:
There is a `stack-use-after-scope` (reported by ASAN, also crashes in opt mode) with `function_ref`. The callback in the greedy pattern rewriter is defined inside of an `if` check:
```c++
    function_ref<bool(const Pattern &)> canApply = {};
    function_ref<void(const Pattern &)> onFailure = {};
    function_ref<LogicalResult(const Pattern &)> onSuccess = {};
    bool debugBuild = false;
#ifdef NDEBUG
    debugBuild = true;
#endif // NDEBUG
    if (debugBuild || config.listener) {
      canApply = [&](const Pattern &pattern) {
        if (this->config.listener) { /* ... */ }
        // ...
      }
      // ...
    }

    // `canApply` points to a lambda that is out of scope.
    LogicalResult matchResult =
        matcher.matchAndRewrite(op, *this, canApply, onFailure, onSuccess);
```

`function_ref` is a "non-owning function wrapper", but the lambda captures `this`.

Changing to `std::function` is one way to fix it. I could also just always pass a lambda. That would actually be my preferred solution, but there is a slight overhead when running in opt mode and without listener because the callback would always be called (even if it does not do anything):
```c++
    LogicalResult matchResult = matcher.matchAndRewrite(
        op, *this,
        /*canApply=*/[&](const Pattern &pattern) {
          if (this->listener) { /* ... */ }
          // ...
        },
        /*onFailure=*/[&](const Pattern &pattern) { /* ... */},
        /*onSuccess=*/[&](const Pattern &pattern) { /* ... */});
```

What do you think?

Reply:
(The existing code seemed to care about performance here; There were different `canApply` etc. depending on `NDEBUG`.)

Reply:
That can explain why you changed it at the call-site, but I'm puzzled about this function: it does not capture the callback as far as I can tell.

Reply:
> What are you referring to with this function?

Where this comment thread is anchored: `matchAndRewrite`

> The problem here is really just caused by the fact that the canApply = assignment is inside of a nested scope. And the lambda object is dead by the time matcher.matchAndRewrite is called. I.e., the canApply function_ref points to an already free'd lambda. At least that's my understanding.

Yes, but that's a problem for the call-site, I don't quite see where you make the connection to the signature of `matchAndRewrite`?

Reply:
That's a good way to think about it.